---
title: "Reproduction of Bakhoum et al 2018 Bulk Transcriptomics"
author: "Pedro Batista Tan"
date: "`r strftime(Sys.Date(), '%B %d %Y')`"
output:
  html_document:
    toc: yes
    number_sections: yes
  pdf_document:
    toc: yes
---
# Install/load packages
```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, warning=FALSE, message = FALSE, dpi = 300, error = FALSE
  # fig.width = 10, fig.height = 6
)

# install.packages("BiocManager")
# BiocManager::install("DESeq2")
# BiocManager::install("apeglm")
library(DESeq2)
# install.packages("broom")
library(broom)
# install.packages("patchwork")
library(patchwork)
# install.packages("ggfortify")
library(ggfortify)
# install.packages("colorspace")
library(colorspace)
# install.packages("GGally")
library(GGally)
# install.packages("ggrepel")
library(ggrepel)
# install.packages("AnnotationDbi")
library(AnnotationDbi)
# install.packages("tidyverse")
library(tidyverse)
# install.packages("gplots")
library(gplots)
# install.packages("pheatmap")
library(pheatmap)
# install.packages("RColorBrewer")
library(RColorBrewer)
#BiocManager::install("org.Hs.eg.db")
library(org.Hs.eg.db)
#install.packages("msigdbr")
library(msigdbr)
# BiocManager::install("fgsea")
library(fgsea)
```

# Load data
```{r load_data, include=FALSE}
# Load transcriptomics count data
# Each row is a gene, each column a sample
df.counts.raw <- read.csv("data/GSE98183_Counts.geneSymbols.75bpPE.csv") 

# load sample design
design <- read.csv2("data/samples.csv")
colnames(design) <- c("Sample", "Name", "Group")

# Sample names
colnames(df.counts.raw)
colnames(df.counts.raw) <- c("gene", "MK", "MKH", "Kb", "Ka", "cont")

```

```{r}
# Convert dataframe to long format
df.counts.long <- df.counts.raw %>% pivot_longer(cols = c(-gene), values_to = "Count",  names_to = "Sample")
```

```{r}
# Some of the gene names in the count data correspond to the gene_name, some to the ENSEMBL gene_id

# Create a "gene" column in the ensembl_db that corresponds to either the id or the name, depending on how it was annotated in the count data


# gene_ensembl_db <- df.counts.raw %>% select(gene) %>% inner_join(ensembl_db %>% mutate(gene = gene_name)) %>% rbind(
#   df.counts.raw %>% select(gene) %>% inner_join(ensembl_db %>% mutate("gene" = gene_id))
#   )
```
# Filtering out genes with 0 counts
```{r}
# filter out genes for which all the samples have 0 counts.
# Also create a column for log2 of pseudocounts
df.counts.f <- df.counts.long %>%
  group_by(gene) %>% filter(any(Count > 0)) %>% 
  mutate(counts_log2 = log2(Count+1)) %>% ungroup()

```

# Differential Expression

## Preparing the data for DESeq2
```{r}
# join design data
# Convert CIN-medium label to CIN-High, as done in the paper
design_c <- design %>% mutate(Group = ifelse(Group == "CIN-medium", "CIN-high", Group))

df.counts.f <- df.counts.f %>% left_join(design_c,  by = "Sample")
```

Create the object for countData. Transform the df.diff to wide format, set the genes as row names and convert to matrix.

```{r}
# select only the necessary columns
# tranform to wide format
# set the row names
# transform to matrix.
df.diff.wide <- df.counts.f %>% 
  arrange(Group) %>% 
  select(gene, Count, Sample) %>% 
  pivot_wider(
    names_from = Sample, 
    values_from =  Count
    ) %>% column_to_rownames("gene") %>%
  as.matrix()

head(df.diff.wide)
```

```{r}
  # get the unique combinations of sample and group, arrange by group
  # convert Sample and group columns to factors.
  # Rows must be in the same order as the columns of the count matrix
df.diff.design <- df.counts.f %>% 
  distinct(Sample, Group) %>%
  arrange(Group) %>%
  mutate(
    Sample = as.factor(Sample),
    Group = as.factor(Group)
  )

df.diff.design
```

Create the DESeq2 object and estimate size factors to apply counts normalization

```{r}
# create the deseq dataset
deseq.dataset <- DESeqDataSetFromMatrix(
  countData=as.matrix(df.diff.wide),
  colData=df.diff.design,
  design=~Group
) %>%
  estimateSizeFactors()

# Change factor levels to compare CIN-High x CIN-low (default is alphabetical)
deseq.dataset$Group <- factor(deseq.dataset$Group, levels = c("CIN-low","CIN-high"))

```

## Dispersion estimation
```{r}
# create DESeq object
dds <- DESeq(deseq.dataset)
```

```{r}
resultsNames(dds)
res <- results(dds)

summary(res)
```

```{r}
plotDispEsts(dds, log = "xy")
```

## Differential expression
```{r}
#?lfcShrink
resLFC <- lfcShrink(dds, coef = "Group_CIN.high_vs_CIN.low", type = "apeglm")

df.deseq.results.noLFC <- res %>%
  as.data.frame() %>%
  rownames_to_column("gene")

df.deseq.results <- resLFC %>%
  as.data.frame() %>%
  rownames_to_column("gene")

df.deseq.results %>% 
  head()

# df.deseq.results2 <- dds %>%
#   lfcShrink(contrast=c("Group","CIN-high","CIN-low"), type = "normal") %>%
#   as.data.frame() %>%
#   rownames_to_column("gene")
```

The results can be plotted in an MA plot:

```{r}
res %>% 
  plotMA(ylim=c(-5,5))

# MA plot on shrunken Log2 Fold Changes, also removes the ones with low counts
resLFC  %>% 
  plotMA(ylim=c(-5,5))
```

Draw an histogram of the p-values and the adjusted p-values

```{r}
# histogram of pvalues
df.deseq.results %>% 
  ggplot(aes(pvalue)) + geom_histogram() + labs(
    title = "DESeq2 p-value distribution", 
    x = "DESeq2 P-value", 
    y = "Number of genes"
  ) +  # you can add two ggplot objects with `+` with the patchwork package
  df.deseq.results %>% 
  ggplot(aes(padj)) + geom_histogram() + labs(
    title = "DESeq2 Adjusted p-value distribution", 
    x = "DESeq2 adjusted P-value", 
    y = "Number of genes"
  )
 # histogram of adj pvalues
```

## Volcano plot
```{r}
alpha = 1e-15
log2FCthresh = 3
df.deseq.results <- df.deseq.results %>% mutate(
  within_threshold = (padj < alpha & abs(log2FoldChange) > log2FCthresh)
)

#df.volcano <- df.deseq.results %>% mutate(threshold = (abs(log2FoldChange) > 2 & padj < alpha))

df.deseq.results %>% ggplot(
  aes(x=log2FoldChange , y= -log10(padj + 1e-200), col = within_threshold)) +     geom_point(alpha = 0.5) +
geom_vline(xintercept = 0, col = "black") + 
geom_vline(xintercept = c(-log2FCthresh, log2FCthresh), col = "red") +   geom_hline(yintercept = -log10(alpha), colour = "red") + 
geom_text_repel(
  aes(x = log2FoldChange, y = -log10(padj + 1e-200), label = gene), data = df.deseq.results %>% filter(within_threshold), colour = "grey20", min.segment.length = 0
  ) +
  theme(legend.position = "none", plot.title = element_text(hjust= 0.5)) +
  labs(
    title = "Volcano plot", 
    x = "Effect size: log2(fold-change)", 
    y = "-log10(adjusted p-value)"
  ) + ylim(0, 50) + 
  scale_x_continuous(breaks = seq(-6,9, by = 1))

```
```{r}
  # filter
df.deseq.results %>%
  filter(within_threshold == TRUE) %>% nrow()

df.deseq.results %>%
  filter(within_threshold == TRUE) %>%
  arrange(-log2FoldChange, padj) %>% head()
```
# Perform PCA on VST data
```{r}
# Apply a variance stabilizing transform
vsd <- vst(dds, blind=FALSE)
# vsd %>% assay() %>% head()

plotPCA(vsd, intgroup=c("Group")) +
geom_text_repel(
  aes(x = PC1, y = PC2, label = name), data = plotPCA(vsd, intgroup=c("Group"), returnData = TRUE))
```
# Load msigdb hallmarks gene set
```{r}
all_gene_sets <- msigdbr("Homo sapiens")

msigdb_hallmarks_set <- filter(all_gene_sets, gs_cat == "H") %>% select(gs_name, gene_symbol)

msigdb_hallmarks_set <- split(msigdb_hallmarks_set$gene_symbol, msigdb_hallmarks_set$gs_name)

msigdb_hallmarks_names <- read.csv2("data/names_hallmarks.csv")

rm(all_gene_sets)
# Another option is loading files with gene sets
# msigdb_hallmarks <- gmtPathways("data/h.all.v7.4.symbols.gmt")
# msigdb_hallmarks %>% head() %>% lapply(head)

```

# FGSEA Shrunk LFC
```{r}
CIN_fc_rnk <- df.deseq.results %>% select(gene, log2FoldChange) %>% deframe()

fgseaRes <- fgsea(pathways = msigdb_hallmarks_set, 
                  stats    = CIN_fc_rnk,
                  minSize  = 15,
                  maxSize  = 500)

fgseaRes <- fgseaRes %>% mutate(within_threshold = pval < 0.05, "Enriched_in" = ifelse(NES > 0, "CIN-high", "CIN-low")) %>% left_join(msigdb_hallmarks_names)


fgseaRes %>% mutate(pval = ifelse(pval < 0.00001, 0.0000101, pval)) %>% ggplot(aes(x = log10(pval), y = NES)) +
  geom_point(aes(size = 0.5, alpha = 0.3, col = Enriched_in)) + 
  geom_hline(yintercept = 0, col = "black") +
  geom_text_repel(
  aes(x = log10(pval), y = NES, label = name), data = fgseaRes %>% filter(within_threshold), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0
  ) +
  labs(x = "log10(pval)", y = "Normalized Enrichment Score", col = "Enriched in") + 
  scale_y_continuous(limits = c(-2,2), breaks = seq(-2,2,1)) + 
  scale_x_continuous(trans = "reverse", limits = c(0, -5), breaks = c(-5, -4, -3, -2, -1, 0) , labels = c("<0.00001", "0.0001", "0.001", "0.01", "0.1", "1")) + scale_fill_manual(values = c("red","blue")) + guides(alpha = FALSE, size = FALSE)

```

# FGSEA not shrunk LFC
```{r}
CIN_fc_rnk <- df.deseq.results.noLFC %>% select(gene, log2FoldChange) %>% deframe()

fgseaRes.noLFC <- fgsea(pathways = msigdb_hallmarks_set, 
                  stats    = CIN_fc_rnk,
                  minSize  = 15,
                  maxSize  = 500)

fgseaRes.noLFC <- fgseaRes.noLFC %>% mutate(within_threshold = pval < 0.05, "Enriched_in" = ifelse(NES > 0, "CIN-high", "CIN-low")) %>% left_join(msigdb_hallmarks_names)

# plotEnrichment(msigdb_hallmarks_set[["HALLMARK_KRAS_SIGNALING_UP"]],
#                CIN_fc_rnk) + labs(title="KRAS Signaling")
# 
# plotEnrichment(msigdb_hallmarks_set[["HALLMARK_INFLAMMATORY_RESPONSE"]],
#                CIN_fc_rnk) + labs(title="Inflammatory Response")
# 
# plotEnrichment(msigdb_hallmarks_set[["HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION"]],
#                CIN_fc_rnk) + labs(title="EMT")
# 

fgseaRes.noLFC %>% ggplot(aes(x = log10(pval), y = NES)) +
  geom_point(aes(size = 0.5, alpha = 0.3, col = Enriched_in)) + 
  geom_hline(yintercept = 0, col = "black") +
  geom_text_repel(
  aes(x = log10(pval), y = NES, label = name), data = fgseaRes.noLFC %>% filter(within_threshold), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0
  ) +
  labs(x = "log10(pval)", y = "Normalized Enrichment Score", col = "Enriched in") + 
  scale_y_continuous(limits = c(-3,3), breaks = seq(-3,3,1)) + 
  scale_x_continuous(trans = "reverse") + scale_fill_manual(values = c("red","blue")) + guides(alpha = FALSE, size = FALSE)

```

# GSEA
Run on Broad institute GSEA application
```{r}
# GSEA results
GSEA_pos <- read.table("data/CIN_high_low_preranked_log2FC.GseaPreranked/gsea_report_for_na_pos_1618265603872.tsv", header = TRUE, sep = "\t")

GSEA_neg <- read.table("data/CIN_high_low_preranked_log2FC.GseaPreranked/gsea_report_for_na_neg_1618265603872.tsv", header = TRUE, sep = "\t")

GSEA_res <- rbind(GSEA_pos, GSEA_neg) %>% mutate(within_threshold = FDR.q.val < 0.3) %>% mutate("Enriched_in" = ifelse(NES > 0, "CIN-high", "CIN-low"))

# GSEA_res %>% select(-c(X, within_threshold)) %>% arrange(-NES) %>%  write_csv2("data/GSEA_CIN_DE_hallmarks.csv")

GSEA_res <- GSEA_res %>% left_join(msigdb_hallmarks_names %>% rename(NAME = pathway))
		
GSEA_res %>% 
  ggplot(aes(x = log10(FDR.q.val), y = NES)) +
  geom_point(aes(size = 0.5, alpha = 0.3, col = Enriched_in)) + 
  geom_hline(yintercept = 0, col = "black") +
  geom_text_repel(
  aes(x = log10(FDR.q.val), y = NES, label = name), data = GSEA_res %>% filter(within_threshold), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0
  ) +
  labs(x = "log10(FDRq)", y = "Normalized Enrichment Score", col = "Enriched in") + 
  scale_y_continuous(limits = c(-2,2), breaks = seq(-2,2,1)) + 
  scale_x_continuous(trans = "reverse", limits = c(0, -3), breaks = c(-3, -2, -1, 0) , labels = c("<0.001", "0.01", "0.1", "1")) + scale_fill_manual(values = c("red","blue")) + guides(alpha = FALSE, size = FALSE)

```

# Compare with original GSEA
```{r}
Bulk_gsea <- read.csv2("data/Supp_table6.csv")

# GSEA x Bulk
GSEA_comp <- Bulk_gsea %>% select(NAME, NES, FDR.q.val) %>% rename(orig_NES = NES, orig_FDR.q.val = FDR.q.val) %>% left_join(
  GSEA_res %>% select(NAME, NES, FDR.q.val)) %>% left_join(
  msigdb_hallmarks_names %>% rename(NAME = pathway)
  ) %>% mutate(sign_agreement = sign(orig_NES) == sign(NES))

GSEA_comp %>% ggplot(aes(x = orig_NES, y = NES, col = log(FDR.q.val))) + 
  geom_point() + geom_abline(slope = 1) + 
  geom_text_repel(
    aes(x = orig_NES, y = NES, label = name), data = GSEA_comp %>% filter(!sign_agreement), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0, max.overlaps = 20
  ) + labs(x = "Orig GSEA NES", y = "GSEA NES") +
  scale_color_gradient2(low="black", mid = "blue", high="red", midpoint = -2)

```
# Compare FGSEA and original GSEA NES
```{r}
# fGSEA x original GSEA
fGSEA_comp <- Bulk_gsea %>% select(NAME, NES, FDR.q.val) %>% rename(orig_NES = NES, orig_FDR.q.val = FDR.q.val) %>% left_join(
  fgseaRes %>% select(pathway, NES, pval) %>% rename(NAME = pathway)
  ) %>% 
  left_join(
  msigdb_hallmarks_names %>% rename(NAME = pathway)
  ) %>% mutate(sign_agreement = sign(orig_NES) == sign(NES))

fGSEA_comp %>% ggplot(aes(x = orig_NES, y = NES, col = log(pval))) + 
  geom_point() + geom_abline(slope = 1) + 
  geom_text_repel(
    aes(x = orig_NES, y = NES, label = name), data = fGSEA_comp %>% filter(!sign_agreement), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0, max.overlaps = 20
  ) + labs(x = "Orig GSEA NES", y = "fGSEA NES") +
  scale_color_gradient2(low="black", mid = "blue", high="red", midpoint = -3)

```
# Compare not shrunk FGSEA and original GSEA NES
```{r}
# fGSEA x Bulk
fGSEA_comp <- Bulk_gsea %>% select(NAME, NES, FDR.q.val) %>% rename(orig_NES = NES, orig_FDR.q.val = FDR.q.val) %>% left_join(
  fgseaRes.noLFC %>% select(pathway, NES, pval) %>% rename(NAME = pathway)
  ) %>% 
  left_join(
  msigdb_hallmarks_names %>% rename(NAME = pathway)
  ) %>% mutate(sign_agreement = sign(orig_NES) == sign(NES))

fGSEA_comp %>% ggplot(aes(x = orig_NES, y = NES, col = log(pval))) + 
  geom_point() + geom_abline(slope = 1) + 
  geom_text_repel(
    aes(x = orig_NES, y = NES, label = name), data = fGSEA_comp %>% filter(!sign_agreement), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0, max.overlaps = 20
  ) + labs(x = "Orig GSEA NES", y = "fGSEA NES") +
  scale_color_gradient2(low="black", mid = "blue", high="red", midpoint = -3)

```



