---
title: "Bakhoum et al 2018 scRNA-seq"
author: "Pedro Batista Tan"
date: "`r strftime(Sys.Date(), '%B %d %Y')`"
output: 
  html_document:
    toc: true
    number_sections: true
---
# Loading Packages
```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, warning=FALSE, message = FALSE, dpi = 300
  # fig.width = 10, fig.height = 6, output: html_notebook
)

# install.packages("BiocManager") #(only if you do not have BiocManager)
#BiocManager::install("fgsea", "GO.db", "GOstats", "org.Hs.eg.db")

# install.packages("Seurat", "patchwork", "ggrepel", "sparseMatrixStats", "msigdbr", "tidyverse", "proxy", "plotly", "pheatmap", "gplots")

library(Seurat)
library(patchwork)
library(ggrepel)
library(sparseMatrixStats)
library(msigdbr)
library(GO.db)
library(GOstats)
library(org.Hs.eg.db)
library(tidyverse)
library(proxy)
library(fgsea)
library(plotly)
library(pheatmap)
library(gplots)

set.seed(42)
select = dplyr::select
rename = dplyr::rename
```

# Loading data
```{r}
# Load filtered and combined data
scRNA_c <- readRDS(file = "data/scRNA_c.rds")

export = FALSE

# generate output dir path named data
source.output.dir <- file.path("output_sc")

# if source output dir does not exist, create it
if (!dir.exists(source.output.dir)) {
    dir.create(source.output.dir)
} else{
  print("Output folder already exists")
}

```

```{r}
mat <- Seurat::GetAssayData(scRNA_c, assay = "RNA", slot = "data")
mat_counts <- Seurat::GetAssayData(scRNA_c, assay = "RNA", slot = "counts")

# Compare counts and data from the first 10 cells
df1 <- mat[, 1:10] %>% as.data.frame()
df2 <- mat_counts[, 1:10] %>% as.data.frame()

df1 %>% rownames_to_column("gene") %>% pivot_longer(cols = c(-gene), values_to = "data", names_to = "sample") %>% left_join(
df2 %>% rownames_to_column("gene") %>% pivot_longer(cols = c(-gene), values_to = "count", names_to = "sample")
) %>% ggplot(aes(x = count, y = data, col = sample)) + geom_point()

```

```{r}
median_library_size <- scRNA_c@assays$RNA %>% colSums() %>% median()

# Normalize dataset
scRNA_cn <- NormalizeData(scRNA_c, normalization.method = "LogNormalize", scale.factor = median_library_size)

mat <- Seurat::GetAssayData(scRNA_cn, assay = "RNA", slot = "data")
mat_counts <- Seurat::GetAssayData(scRNA_cn, assay = "RNA", slot = "counts")

# Compare counts and data from the first 10 cells
df1 <- mat[, 1:10] %>% as.data.frame()
df2 <- mat_counts[, 1:10] %>% as.data.frame()

df1 %>% rownames_to_column("gene") %>% pivot_longer(cols = c(-gene), values_to = "data", names_to = "sample") %>% left_join(
df2 %>% rownames_to_column("gene") %>% pivot_longer(cols = c(-gene), values_to = "count", names_to = "sample")
) %>% ggplot(aes(x = count, y = data, col = sample)) + geom_point()

```

# Normalize and scale
```{r}
median_library_size <- scRNA_c@assays$RNA %>% colSums() %>% median()

# Normalize dataset
scRNA_cn <- NormalizeData(scRNA_c, normalization.method = "LogNormalize", scale.factor = median_library_size)

# find highly variable features with a variance stabilizing transform
scRNA_cn <- FindVariableFeatures(scRNA_cn, selection.method = "vst", nfeatures = 2000)

VariableFeaturePlot(scRNA_cn) %>% LabelPoints(points = head(VariableFeatures(scRNA_cn), 30), repel = TRUE)

# Scale dataset
scRNA_cn <- ScaleData(scRNA_cn, features = rownames(scRNA_cn))

# Add CIN status meta data
scRNA_cn@meta.data <- scRNA_cn@meta.data %>% mutate(CIN_status = ifelse(orig.ident %in% c("dnMCAK"),"CIN-high","CIN-low"))

rm(scRNA_c)
```

# Perform PCA
```{r}
# Perform PCA
scRNA_cn <- RunPCA(scRNA_cn, features = VariableFeatures(object = scRNA_cn), npcs = 200)

# Plot PCA results
Idents(scRNA_cn) <- "orig.ident"
VizDimLoadings(scRNA_cn, dims = 1:2, reduction = "pca")
DimPlot(scRNA_cn, reduction = "pca")
ElbowPlot(scRNA_cn, ndims = 200)

DimHeatmap(scRNA_cn, dims = 1:10, cells = 500, balanced = TRUE)

# Check the total variance explained
# On Seurat 3:
pca <- scRNA_cn[["pca"]]
# Get the total variance of the scaled counts matrix:
mat <- Seurat::GetAssayData(scRNA_cn, assay = "RNA", slot = "scale.data")
mat <- mat[rownames(mat) %in% VariableFeatures(scRNA_cn), ]

total_variance <- sum(matrixStats::rowVars(mat))

eigValues = (pca@stdev)^2  ## EigenValues
varExplained = eigValues / total_variance

varExplained

cumsum(varExplained) %>% as.data.frame() %>% rename("cumvar" = ".") %>% ggplot(aes(x = c(1:200), y = cumvar)) + geom_point() + labs(x = "PC #", y = "Cumulative Variance")

rm(mat)
rm(pca)
#Perform PCA "manually" on the matrix to check Seurat % variance explained

# USV <- svd(mat)	
# ssqtotal <- sum(mat*mat)
# 
# T <- USV$u %*% diag(USV$d)
# head(T)
# P <- USV$v
# npc <- 200
# ssqtot <- 0 * (1:npc)
# for (i in 1:npc){
#   print(i)
#   Xest  <- T[,1:i] %*% t(P[,1:i])
#   ssqtot[i] <- 100 * (1-sum((mat-Xest)*(mat-Xest))/ssqtotal)
# }
# 
# ssq=c(ssqtot[1],ssqtot[2:10]-ssqtot[1:9])
# rawPCA <- data.frame(ssq=ssq,ssqtot=ssqtot)
# 
# rm(T)
# rm(P)
# rm(rawPCA)
# rm(Xest)
# rm(USV)

```

```{r}
# Compute a null distribution for feature scores by permuting 1% of the data multiple times and compare this with information within each PC to obtain a significance estimate
scRNA_cn <- JackStraw(scRNA_cn, num.replicate = 100, dims = 50)
scRNA_cn <- ScoreJackStraw(scRNA_cn, dims = 1:50)
JackStrawPlot(scRNA_cn, dims = 1:50)
# 1:50
```
# Cell cycle assignment
```{r}
# A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.  We can
# segregate this list into markers of G2/M phase and markers of S phase
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

# Assign and inspect cell cycle scores
scRNA_cn <- CellCycleScoring(scRNA_cn, s.features = s.genes, g2m.features = g2m.genes)

scRNA_cn[[]] %>% select(S.Score, G2M.Score, Phase) %>% mutate_if(is.character, as.factor) %>% summary()

Idents(scRNA_cn) <- "Phase"
RidgePlot(scRNA_cn, features = c("PCNA", "TOP2A", "MCM6", "MKI67"), ncol = 2)
Idents(scRNA_cn) <- "orig.ident"
RidgePlot(scRNA_cn, features = c("PCNA", "TOP2A", "MCM6", "MKI67"), ncol = 2)

```

# Run UMAP and TSNE
```{r}
# Dimensionality reduction with UMAP on variable features instead of PCs
scRNA_cn <- RunUMAP(scRNA_cn, reduction.use = NULL, features = VariableFeatures(scRNA_cn))

# Store this UMAP as a custom reduction
scRNA_cn[["UMAP_features"]] <- CreateDimReducObject(embeddings = scRNA_cn@reductions$umap@cell.embeddings, key = "UMAP_features", assay = DefaultAssay(scRNA_cn))

# Run UMAP and project in 3D dimensions
scRNA_cn <- RunUMAP(scRNA_cn, dims = 1:200, n.components = 3L)

# Store this UMAP as a custom reduction
scRNA_cn[["umap3d"]] <- CreateDimReducObject(embeddings = scRNA_cn@reductions$umap@cell.embeddings, key = "umap3d", assay = DefaultAssay(scRNA_cn))

# Dimensionality reduction with UMAP and tSNE on PCs
scRNA_cn <- RunUMAP(scRNA_cn, dims = 1:200)

scRNA_cn <- RunTSNE(scRNA_cn, dims.use = 1:200, reduction.use = "pca", perplexity = 30)

```

# Clustering cells
```{r}
scRNA_cn <- FindNeighbors(scRNA_cn, features = VariableFeatures(object = scRNA_cn))
scRNA_cn <- FindClusters(scRNA_cn, resolution = 0.5)

scRNA_cn <- StashIdent(scRNA_cn, save.name = "clusterID")

# Inspect cluster information
#head(Idents(scRNA_cn), 5)
Idents(scRNA_cn) <- "clusterID"

table(scRNA_cn$orig.ident)
table(scRNA_cn$clusterID)
prop.table(table(scRNA_cn$clusterID))
prop.table(table(scRNA_cn$clusterID, scRNA_cn$orig.ident), margin = 2)
prop.table(table(scRNA_cn$orig.ident, scRNA_cn$clusterID), margin = 2)

prop.table(table(scRNA_cn$CIN_status))
prop.table(table(scRNA_cn$clusterID, scRNA_cn$CIN_status), margin = 2)
prop.table(table(scRNA_cn$CIN_status, scRNA_cn$clusterID), margin = 2)
prop.table(table(scRNA_cn$Phase, scRNA_cn$CIN_status), margin = 2)

table(scRNA_cn$clusterID, scRNA_cn$CIN_status) %>% as.data.frame() %>% rename("clusterID" = Var1, "group" = Var2) %>% ggplot(aes(x = clusterID, y = group, fill = Freq)) + geom_tile() + scale_fill_distiller(palette = "PuBu") + geom_text(aes(label=Freq)) + labs(y = "", fill = "Number of cells")

```

# Visualization
```{r}
DimPlot(scRNA_cn, reduction = "UMAP_features", group.by = "CIN_status")
DimPlot(scRNA_cn, reduction = "UMAP_features", group.by = "orig.ident")
DimPlot(scRNA_cn, reduction = "UMAP_features", group.by = "Phase")
DimPlot(scRNA_cn, reduction = "UMAP_features", group.by = "clusterID")

DimPlot(scRNA_cn, reduction = "umap", group.by = "CIN_status")
DimPlot(scRNA_cn, reduction = "umap", group.by = "orig.ident")
DimPlot(scRNA_cn, reduction = "umap", group.by = "Phase")
DimPlot(scRNA_cn, reduction = "umap", group.by = "clusterID")

DimPlot(scRNA_cn, reduction = "tsne", group.by = "CIN_status")
DimPlot(scRNA_cn, reduction = "tsne", group.by = "orig.ident")
DimPlot(scRNA_cn, reduction = "tsne", group.by = "Phase")
DimPlot(scRNA_cn, reduction = "tsne", group.by = "clusterID")

```
# 3D UMAP
```{r}
# Extract cell embedding information from Seurat Object
# umap_1 <- scRNA_cn[["umap3d"]]@cell.embeddings[,1]
# umap_2 <- scRNA_cn[["umap3d"]]@cell.embeddings[,2]
# umap_3 <- scRNA_cn[["umap3d"]]@cell.embeddings[,3]

# Visualize what headings are called so that you can extract them to form a dataframe
Embeddings(object = scRNA_cn, reduction = "umap") %>% head()
Embeddings(object = scRNA_cn, reduction = "umap3d") %>% head()

# Prepare a dataframe for cell plotting
plot.data <- FetchData(object = scRNA_cn, vars = c("umap3d_1", "umap3d_2", "umap3d_3", "seurat_clusters", "Phase", "orig.ident"))

# Make a column of row name identities (these will be your cell/barcode names)
plot.data$label <- paste(rownames(plot.data))

# Plot your data
#When you visualize your plotly object, hovering your mouse pointer over a point shows cell names

fig <- plot_ly(data = plot.data, 
        x = ~umap3d_1, y = ~umap3d_2, z = ~umap3d_3, 
        color = ~seurat_clusters, 
        colors = c("lightseagreen", "gray50", "darkgreen", "red4", "red", "turquoise4", "black", "yellow4", "royalblue1", "lightcyan3", "peachpuff3",
"khaki3", "gray20", "orange2", "royalblue4", "yellow3", "gray80", "darkorchid1", "lawngreen", "plum2", "darkmagenta"),
        type = "scatter3d", 
        mode = "markers", 
        marker = list(size = 3, width=2), # controls size of points
        text=~orig.ident,
        hoverinfo="text") 
fig

fig <- plot_ly(data = plot.data, 
        x = ~umap3d_1, y = ~umap3d_2, z = ~umap3d_3, 
        color = ~Phase,
        type = "scatter3d", 
        mode = "markers", 
        marker = list(size = 3, width=2), # controls size of points
        text=~orig.ident,
        hoverinfo="text") 

fig

# Before you plot, set the ranges of the axis you desire. This set axis range will be 
# present across all clusters, and plotly will not adjust for axis length anymore
# this axis length will persist even when selecting some clusters

# xaxis
# axx <- list(nticks = 4,  range = c(-10,10))

# yaxis
# axy <- list(nticks = 4, range = c(-10,10))

#zaxis
# axz <- list(nticks = 4, range = c(-10,10))


# fig <- fig %>% layout(scene = list(xaxis=axx,yaxis=axy,zaxis=axz))
# fig_cube <- fig %>% layout(scene = list(xaxis=axx,yaxis=axy,zaxis=axz, aspectmode='cube')) # To maintain cubic aspect

rm(plot.data)
```


```{r}
# Say you wanto make a gene-expression 3D plot, where you can plot gene expression against a color scale
# Here using the same seurat object as above, we extract gene expression information for beta-actin 'ACTB'
# Here we concentrate on SCT normalized data, or log normalized RNA NOT raw counts.
# In addition if you want, you may look at normalised-RNA, SCT or integrated slots, to look at gene expression
# Setting your DefaultAssay() will inform R which assay to pick up expression data from.

# DefaultAssay(object = yourseuratobject)
# DefaultAssay(object = yourseuratobject) <- "RNA"
# DefaultAssay(object = yourseuratobject) <- "integrated"
# DefaultAssay(object = yourseuratobject) <- "SCT"

# create a dataframe
# plot.data <- FetchData(object = yourseuratobject, vars = c("UMAP_1", "UMAP_2", "UMAP_3", "ACTB"), slot = 'data')

# Say you want change the scale, so that every cell having an expression >1 will be one color
# Basically, you are re-adjusting the scale here, so that any cell having a certain expression will light up on your 3D plot

# First make another column in your dataframe, where all values above 1 are re-assigned a value of 1
# This information is stored in the 'changed' column of your dataframe

# plot.data$changed <- ifelse(test = plot.data$ACTB <1, yes = plot.data$ACTB, no = 1)

# Add the label column, so that now the column has 'cellname-its expression value'

# plot.data$label <- paste(rownames(plot.data)," - ", plot.data$ACTB, sep="")

# Plot your data, in this example my Seurat object had 21 clusters (0-20), and cells express a gene called ACTB

# you can just run this against the column for the gene as well using ~ACTB, the algorith will automatically scale in that case based on maximal and minimal values
# plot_ly(data = plot.data, 
#         x = ~UMAP_1, y = ~UMAP_2, z = ~UMAP_3, 
#         color = ~changed, 
#         opacity = .5,
#         colors = c('darkgreen', 'red'), 
#         type = "scatter3d", 
#         mode = "markers",
#         marker = list(size = 5, width=2), 
#         text=~label,
#         hoverinfo="text"
# )

# On running this code the HTML output should appear in RStudio. You can save the output as a
# HTML file. Once you have saved, just open the HTML file in any web browser (double click on the html- file
# and if asked select to open with any web browser like google chrome/safari/mozilla/explorer etc).
# It should be have all of the integrated features you saw in the RStudio output file.
```

# Cluster markers
```{r} 
# Find cluster markers 
Idents(scRNA_cn) <- "clusterID"
cluster.DE <- FindAllMarkers(scRNA_cn, min.pct = 0.05, logfc.threshold = 0)

# Run later overnight
cluster.DE2 <- FindAllMarkers(scRNA_cn, min.pct = 0.05, logfc.threshold = 0, pseudocount.use = 0.001)

#cluster.markers
clustertop10 <- cluster.DE %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)

cluster.DE %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
```
```{r}
DoHeatmap(scRNA_cn, features = clustertop10$gene) + NoLegend()
```

# Plotting signatures
```{r}
# Plotting EMT signature 
EMT_signature <- c("EZH", "JUN", "VIM", "STEAP1", "SOX4", "MMP14", "SHH", "TIMP1", "ZEB1", "ZEB2", "SNAI2")

EMT_signature <- EMT_signature[EMT_signature %in% rownames(scRNA_cn@assays$RNA)]

Idents(scRNA_cn) <- "CIN_status"
for(i in 1:length(EMT_signature)){
p <- VlnPlot(scRNA_cn, features = EMT_signature[i])
print(p)
}

Idents(scRNA_cn) <- "clusterID"
for(i in 1:length(EMT_signature)){
p <- VlnPlot(scRNA_cn, features = EMT_signature[i])
print(p)
}

```

```{r}
# Plotting Genes enriched in the "M" Population from Bakhoum et al Extended figure 6b 
M_markers = c("ITGB5", "ITGB1", "ITGA5", "ITGA10", "IGFBP4", "FN1", "DSC2", "CXCL1","CTNNB1", "BMP4", "BCL2L1", "VIM", "TIMP1", "TGFBR2", "TGFBI", "TGFB1", "PPARG", "NUPR1", "MSN", "LMCD1", "LEF1", "JAG1")

M_markers <- M_markers[M_markers %in% rownames(scRNA_cn@assays$RNA)]

Idents(scRNA_cn) <- "CIN_status"
for(i in 1:length(M_markers)){
p <- VlnPlot(scRNA_cn, features = M_markers[i])
print(p)
}

Idents(scRNA_cn) <- "clusterID"
for(i in 1:length(M_markers)){
p <- VlnPlot(scRNA_cn, features = M_markers[i])
print(p)
# p <- FeaturePlot(scRNA_cn, features = M_markers[i])
# print(p)
}
```

```{r}
# Plot CIN signature from Bakhoum et al 2018 (Supp. table 5), comparing CIN high and CIN low in bulk data
CIN_signature <- c('PELI2','BMP2','SHH','TNS4','RAB3B','ROBO1','ARHGAP28','CHN2','CST1','F13A1','CPVL','SEMA6D','NHSL2','GTF2IP7','DPYSL3','PCDH7','KHDRBS3','TRAC','TMEM156', 'CST4','CD24','FGF5','NTN4')

CIN_signature <- CIN_signature[CIN_signature %in% rownames(scRNA_cn@assays$RNA)]

Idents(scRNA_cn) <- "CIN_status"

for(i in 1:length(CIN_signature)){
p <- VlnPlot(scRNA_cn, features = CIN_signature[i])
print(p)
}

Idents(scRNA_cn) <- "clusterID"
for(i in 1:length(CIN_signature)){
p <- VlnPlot(scRNA_cn, features = CIN_signature[i])
print(p)
}

```
# Clustering cells on EMT genes
```{r}
colfunc <- colorRampPalette(c("darkblue", "blue", "white", "red", "darkred"))

# Get the total variance of the scaled counts matrix:
mat <- Seurat::GetAssayData(scRNA_cn, assay = "RNA", slot = "scale.data")
mat <- mat[rownames(mat) %in% msigdb_hallmarks_set$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION, ]

EMT_signature %in% msigdb_hallmarks_set$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION
# Z-score matrix counts
gene_means = rowMeans(mat)
gene_SDs = rowSds(mat)

# If the SD is zero, use 1 instead, it won't affect the mean number
# another option is to filter these genes out
gene_SDs[gene_SDs == 0] <- 1

Z_score_mat = (mat - gene_means)/gene_SDs

mat_annotation <- scRNA_cn@meta.data %>% select(orig.ident)

color_annotation <- list(orig.ident =  c("darkslategray3", "blue4", "red"))
names(color_annotation$orig.ident) <- unique(mat_annotation$orig.ident)

heatmap <- pheatmap(Z_score_mat, color = colfunc(100), show_colnames = FALSE, annotation_col = mat_annotation, annotation_colors = color_annotation, main = "pheatmap column scaling")

my.breaks = c(-2.5, 2.5)
my.colors <- c(colorRampPalette(colors = c("blue", "white"))(length(my.breaks)/2), colorRampPalette(colors = c("white", "red"))(length(my.breaks)/2))

heatmap_trim <- pheatmap(Z_score_mat, my.colors, show_colnames = FALSE, annotation_col = mat_annotation, annotation_colors = color_annotation, main = "pheatmap column scaling", breaks = my.breaks)

test <- Z_score_mat %>% t() %>% as.data.frame()
test %>% summary() %>% as.data.frame()
# Using threshold breaks at Z scores of 2.5
Z_score_mat_t <- Z_score_mat
Z_score_mat_t[Z_score_mat_t > 5] <- 5
Z_score_mat_t[Z_score_mat_t < -5] <- -5

heatmap_trim <- pheatmap(Z_score_mat_t, color = colfunc(100), show_colnames = FALSE, annotation_col = mat_annotation, annotation_colors = color_annotation, main = "pheatmap column scaling")

save_pheatmap_png <- function(x, filename, width=3500, height=7000, res = 300) {
  png(filename, width = width, height = height, res = res)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

if(export){
save_pheatmap_png(heatmap, filename = paste0(source.output.dir, "/EMT_heatmap.png"))

save_pheatmap_png(heatmap_trim, filename = paste0(source.output.dir, "/EMT_heatmap_trim.png"))
}

rm(color_annotation)
rm(Z_score_mat_t)
rm(Z_score_mat)
rm(mat)
rm(gene_means)
rm(gene_SDs)
```

# Use only subpopulation M
```{r}
cluster.M <- FindMarkers(scRNA_cn, ident.1 = c(3,4), min.pct = 0.05, logfc.threshold = 0) %>% rownames_to_column("gene") %>% mutate(cluster = "3,4")

cluster.M %>% write.csv2(file = paste0(source.output.dir, "/scRNA_clusterM.csv"), row.names = FALSE)
```

# Export
```{r}
if(export){
saveRDS(scRNA_cn, file = paste0(source.output.dir, "/Bakhoum_scRNA.rds"))
  
cluster.DE %>% write.csv2(file = paste0(source.output.dir, "/Bakhoum_scRNA_cluster_markers.csv"), row.names = FALSE)

cluster.DE2 %>% write.csv2(file = paste0(source.output.dir, "/Bakhoum_scRNA_cluster_markers2.csv"), row.names = FALSE)

# Export data for pre ranked GSEA
# Convert p-values of zero to 1e-300, then calculate the preranking statistic and export
CIN.DE %>% mutate(
  p_val = ifelse(p_val == 0, 1e-300, p_val),
  signlogFC_pval = sign(avg_log2FC) * -log10(p_val)
  ) %>% select(gene, signlogFC_pval) %>% rename("NAME" = gene) %>% write.table(file = paste0(source.output.dir, "/Bakhoum_scRNA_CIN_Highlow_GSEA_pval.rnk")
, sep = "\t", row.names = FALSE, quote = FALSE)

CIN.DE %>% select(gene, avg_log2FC) %>% rename("NAME" = gene) %>% write.table(file = paste0(source.output.dir, "/Bakhoum_scRNA_CIN_Highlow_GSEA_fc.rnk")
, sep = "\t", row.names = FALSE, quote = FALSE)


}

scRNA_cn <- readRDS(file = paste0(source.output.dir, "/Bakhoum_scRNA.rds"))

cluster.DE <- read.csv2(paste0(source.output.dir, "/Bakhoum_scRNA_cluster_markers.csv"))

cluster.DE2 <- read.csv2(file = paste0(source.output.dir, "/Bakhoum_scRNA_cluster_markers2.csv"))

```

# Load msigdb hallmarks gene set
```{r}
all_gene_sets <- msigdbr("Homo sapiens")

msigdb_hallmarks_set <- filter(all_gene_sets, gs_cat == "H") %>% select(gs_name, gene_symbol)

msigdb_hallmarks_set <- split(msigdb_hallmarks_set$gene_symbol, msigdb_hallmarks_set$gs_name)

msigdb_hallmarks_names <- read.csv2("data/names_hallmarks.csv")

rm(all_gene_sets)
# Another option is loading files with gene sets
# msigdb_hallmarks <- gmtPathways("data/h.all.v7.4.symbols.gmt")
# msigdb_hallmarks %>% head() %>% lapply(head)

```

# Population "M" FGSEA
```{r}
M_rnk <- cluster.M %>% select(gene, avg_log2FC) %>% deframe()

M_fgseaRes <- fgsea(pathways = msigdb_hallmarks_set, 
                  stats    = M_rnk,
                  minSize  = 15,
                  maxSize  = 500)

M_fgseaRes <- M_fgseaRes %>% mutate(within_threshold = pval < 0.01, "Enriched_in" = ifelse(NES > 0, "Cluster 3,4", "Other clusters")) %>% left_join(msigdb_hallmarks_names)

plotEnrichment(msigdb_hallmarks_set[["HALLMARK_KRAS_SIGNALING_UP"]],
               M_rnk) + labs(title="KRAS Signaling")

plotEnrichment(msigdb_hallmarks_set[["HALLMARK_INFLAMMATORY_RESPONSE"]],
               M_rnk) + labs(title="Inflammatory Response")

plotEnrichment(msigdb_hallmarks_set[["HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION"]],
               M_rnk) + labs(title="Inflammatory Response")

#fgseaRes %>% select(-leadingEdge) %>% write.csv2("data/fgsea_cin_high_low.csv")

M_fgseaRes %>% mutate(pval = ifelse(pval < 0.00001, 0.0000101, pval)) %>% ggplot(aes(x = log10(pval), y = NES)) +
  geom_point(aes(size = 0.5, alpha = 0.3, col = Enriched_in)) + 
  geom_hline(yintercept = 0, col = "black") +
  geom_text_repel(
  aes(x = log10(pval), y = NES, label = name), data = M_fgseaRes %>% filter(within_threshold), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0, max.overlaps = 30
  ) +
  labs(x = "log10(pval)", y = "Normalized Enrichment Score", col = "Enriched in") + 
  scale_y_continuous(limits = c(-2,2), breaks = seq(-2,2,1)) + 
  scale_x_continuous(trans = "reverse", limits = c(0, -5), breaks = c(-5, -4, -3, -2, -1, 0) , labels = c("<0.00001", "0.0001", "0.001", "0.01", "0.1", "1")) + scale_fill_manual(values = c("red","blue")) + guides(alpha = FALSE, size = FALSE)

# topPathways <- c(fgseaRes[ES > 0][head(order(pval), n=5), pathway],
#                  rev(fgseaRes[ES < 0][head(order(pval), n=5), pathway]))
# 
# plotGseaTable(msigdb_hallmarks_set[topPathways], CIN_fc_rnk, fgseaRes, gseaParam=0.5, render = TRUE)

if(export){
  M_fgseaRes %>% select(-leadingEdge) %>% write.csv2(file = paste0(source.output.dir, "/scRNA_fgsea_M.csv"), row.names = FALSE)
}

```

# FGSEA DE between clusters
```{r}
cluster.DE.l <- cluster.DE %>% select(gene, cluster, avg_log2FC) %>% split(f = cluster.DE$cluster)

cluster.DE.fgsea <- cluster.DE.l %>% lapply(FUN = function(x){
  x <- x %>% select(-cluster) %>% deframe()
  x <- fgsea(pathways = msigdb_hallmarks_set, 
                  stats = x,
                  minSize = 15,
                  maxSize = 500)
})

# Plot results for the GSEA on each cluster 
lapply(seq_along(cluster.DE.fgsea), FUN = function(i){
  cluster.DE.fgsea[[i]] <- cluster.DE.fgsea[[i]] %>% mutate(
    within_threshold = padj < 0.1, 
    padj = ifelse(padj < 0.0001, 0.000101, padj)
    ) %>% left_join(msigdb_hallmarks_names)

  cluster.DE.fgsea[[i]] %>% ggplot(aes(x = log10(padj), y = NES)) +
  geom_point(aes(size = 0.5, alpha = 0.3, col = as.factor(sign(NES)))) + 
  geom_hline(yintercept = 0, col = "black") +
  geom_text_repel(
  aes(x = log10(padj), y = NES, label = name), data = cluster.DE.fgsea[[i]] %>% filter(within_threshold | name %in% c("EMT", "Inflammatory Response", "IFN-a Response", "IFN-y Response", "TNF-a Signaling via NKFB", "IL2-STAT5 Signaling", "KRAS Signaling Up")), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0, max.overlaps = 20
  ) +
  labs(x = "log10(padj)", y = "Normalized Enrichment Score", col = "Enriched in Cluster", title = paste("Cluster: ", i - 1, sep = "")) + 
  scale_y_continuous(limits = c(-3.5,3.5), breaks = seq(-3.5,3.5,0.5)) + 
  scale_x_continuous(trans = "reverse", limits = c(0, -4), breaks = c(-4, -3, -2, -1, 0) , labels = c("<0.0001", "0.001", "0.01", "0.1", "1")) + scale_fill_manual(values = c("red","blue")) + guides(alpha = FALSE, size = FALSE)
})


# Combine summaries from the fGSEA
cluster.DE.fgsea_summary <- cluster.DE.fgsea[[1]] %>% mutate(cluster = 0) %>% filter(pval == "nothing")

for(i in 1:length(cluster.DE.fgsea)){
  print(i)
  cluster.DE.fgsea_summary <- rbind(
    cluster.DE.fgsea_summary, 
    cluster.DE.fgsea[[i]] %>% mutate(cluster = i-1)
    )
}

if(export){
cluster.DE.fgsea_summary %>% select(-leadingEdge) %>% write.csv2(file = paste0(source.output.dir, "/scRNA_cluster_fgsea.csv"), row.names = FALSE)
}

cluster.DE.fgsea_summary <- read.csv2(file = paste0(source.output.dir, "/scRNA_cluster_fgsea.csv"))
cluster.DE.fgsea <- cluster.DE.fgsea_summary %>% split(f = cluster.DE.fgsea_summary$cluster)

```
## Heatmap Cluster NES
```{r}
save_pheatmap_png <- function(x, filename, width=1200, height=1200, res = 300) {
  png(filename, width = width, height = height, res = res)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

colfunc <- colorRampPalette(c("blue", "white", "red"))

heatmap_data <- cluster.DE.fgsea_summary %>% mutate(NES = ifelse(padj < 0.2, NES, 0)) %>% left_join(msigdb_hallmarks_names) %>% select(name, NES, cluster) %>% pivot_wider(values_from = NES, names_from = name) %>% column_to_rownames("cluster") %>% mutate_all(~replace(., is.na(.), 0)) %>% as.matrix() 

heatmap_data %>% pheatmap(color = colfunc(10))
NES_cluster_pheatmap <- heatmap_data %>% pheatmap(color = colfunc(10))

# heatmap_data %>% heatmap.2(col=colfunc(10), dendrogram = "both", trace = "none", srtCol = 75, lwid = c(0.5,2), lhei = c(0.4, 1))

if(export){
save_pheatmap_png(NES_cluster_pheatmap, filename = paste0(source.output.dir, "/NES_clustering_pheatmap.png"), width=2000, height=1200)

# png(filename = paste0(source.output.dir, "/NES_clustering_heatmap.png"), width = 2000, height = 1200, res = 300)
    }
```

```{r}
DimPlot(scRNA_cn, reduction = "UMAP_features", group.by = "orig.ident")
DimPlot(scRNA_cn, reduction = "UMAP_features", group.by = "clusterID")

table(scRNA_cn$clusterID, scRNA_cn$CIN_status) %>% as.data.frame() %>% rename("clusterID" = Var1, "group" = Var2) %>% ggplot(aes(x = clusterID, y = group, fill = Freq)) + geom_tile() + scale_fill_distiller(palette = "PuBu") + geom_text(aes(label=Freq)) + labs(y = "", fill = "Number of cells")

```
## Using 0.001 Pseudocounts
# FGSEA DE between clusters
```{r}
cluster.DE.l2 <- cluster.DE2 %>% select(gene, cluster, avg_log2FC) %>% split(f = cluster.DE$cluster)

cluster.DE.fgsea2 <- cluster.DE.l2 %>% lapply(FUN = function(x){
  x <- x %>% select(-cluster) %>% deframe()
  x <- fgsea(pathways = msigdb_hallmarks_set, 
                  stats = x,
                  minSize = 15,
                  maxSize = 500)
})

# Plot results for the GSEA on each cluster 
lapply(seq_along(cluster.DE.fgsea2), FUN = function(i){
  cluster.DE.fgsea2[[i]] <- cluster.DE.fgsea2[[i]] %>% mutate(
    within_threshold = padj < 0.1, 
    padj = ifelse(padj < 0.0001, 0.000101, padj)
    ) %>% left_join(msigdb_hallmarks_names)

  cluster.DE.fgsea2[[i]] %>% ggplot(aes(x = log10(padj), y = NES)) +
  geom_point(aes(size = 0.5, alpha = 0.3, col = as.factor(sign(NES)))) + 
  geom_hline(yintercept = 0, col = "black") +
  geom_text_repel(
  aes(x = log10(padj), y = NES, label = name), data = cluster.DE.fgsea2[[i]] %>% filter(within_threshold | name %in% c("EMT", "Inflammatory Response", "IFN-a Response", "IFN-y Response", "TNF-a Signaling via NKFB", "IL2-STAT5 Signaling", "KRAS Signaling Up")), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0, max.overlaps = 20
  ) +
  labs(x = "log10(padj)", y = "Normalized Enrichment Score", col = "Enriched in Cluster", title = paste("Cluster: ", i - 1, sep = "")) + 
  scale_y_continuous(limits = c(-3.5,3.5), breaks = seq(-3.5,3.5,0.5)) + 
  scale_x_continuous(trans = "reverse", limits = c(0, -4), breaks = c(-4, -3, -2, -1, 0) , labels = c("<0.0001", "0.001", "0.01", "0.1", "1")) + scale_fill_manual(values = c("red","blue")) + guides(alpha = FALSE, size = FALSE)
})


# Combine summaries from the fGSEA
cluster.DE.fgsea_summary2 <- cluster.DE.fgsea2[[1]] %>% mutate(cluster = 0) %>% filter(pval == "nothing")

for(i in 1:length(cluster.DE.fgsea2)){
  print(i)
  cluster.DE.fgsea_summary2 <- rbind(
    cluster.DE.fgsea_summary2, 
    cluster.DE.fgsea2[[i]] %>% mutate(cluster = i-1)
    )
}

if(export){
cluster.DE.fgsea_summary2 %>% select(-leadingEdge) %>% write.csv2(file = paste0(source.output.dir, "/scRNA_cluster_fgsea2.csv"), row.names = FALSE)
}

cluster.DE.fgsea_summary2 <- read.csv2(file = paste0(source.output.dir, "/scRNA_cluster_fgsea2.csv"))
cluster.DE.fgsea2 <- cluster.DE.fgsea_summary2 %>% split(f = cluster.DE.fgsea_summary2$cluster)

```
## Heatmap Cluster NES
```{r}
colfunc <- colorRampPalette(c("blue", "white", "red"))

heatmap_data2 <- cluster.DE.fgsea_summary2 %>% mutate(NES = ifelse(padj < 0.2, NES, 0)) %>% left_join(msigdb_hallmarks_names) %>% select(name, NES, cluster) %>% pivot_wider(values_from = NES, names_from = name) %>% column_to_rownames("cluster") %>% mutate_all(~replace(., is.na(.), 0)) %>% as.matrix()

heatmap_data2 %>% pheatmap(color = colfunc(10))
NES_cluster_pheatmap2 <- heatmap_data2 %>% pheatmap(color = colfunc(10))


heatmap_data2 %>% heatmap.2(col=colfunc(10), dendrogram = "both", trace = "none", srtCol = 75)


if(export){
save_pheatmap_png(NES_cluster_pheatmap2, filename = paste0(source.output.dir, "/NES_clustering_pheatmap2.png"), width=2000, height=1200)
}



heatmap_data %>% pheatmap(color = colfunc(10))
NES_cluster_pheatmap <- heatmap_data %>% pheatmap(color = colfunc(10))

# heatmap_data %>% heatmap.2(col=colfunc(10), dendrogram = "both", trace = "none", srtCol = 75, lwid = c(0.5,2), lhei = c(0.4, 1))

if(export){
save_pheatmap_png(NES_cluster_pheatmap2, filename = paste0(source.output.dir, "/NES_clustering_pheatmap2.png"), width=2000, height=1200)
}
```

# Entrez ID symbol Mapping
```{r}
# Map symbols to entrezids using Annotation dbi
entrez_dbi_map <- cluster.DE %>% select(gene) %>% unique()

entrez_dbi_map$entrezid <- mapIds(org.Hs.eg.db,
                    keys=entrez_dbi_map$gene,
                    column="ENTREZID",
                    keytype="SYMBOL",
                    multiVals="first")

entrez_dbi_map %>% filter(is.na(entrezid)) %>% n_distinct() 
entrez_dbi_map %>% filter(!is.na(entrezid)) %>% n_distinct()
```

# GO BP Enrichment
```{r}
cluster.DE.l <- cluster.DE %>% select(gene, p_val, cluster, avg_log2FC) %>% split(f = cluster.DE$cluster)

cluster.DE.GObp <- cluster.DE.l %>% lapply(FUN = function(x){
  x <- x %>% left_join(entrez_dbi_map)
  
  selectUp <- (x %>% filter(p_val < 0.001, avg_log2FC > 0.5))$entrezid
  
  print(paste("Number of genes:" , length(selectUp)))
  
  upParams <- new("GOHyperGParams", 
                geneIds = selectUp, 
                universeGeneIds = unique(cluster.DE$entrezid), 
                annotation = "org.Hs.eg.db", 
                ontology = "BP", 
                pvalueCutoff = 0.01,
                conditional = FALSE, testDirection = "over"
                )
  
  x <- hyperGTest(upParams)
})

cluster.DE.GObp %>% lapply(summary)

# Combine summaries from the GO enrichment tests
cluster.DE.GObp_summary <- cluster.DE.GObp[[1]] %>% summary() %>% mutate(cluster = 0) %>% filter(Pvalue == "nothing")

for(i in 1:length(cluster.DE.GObp)){
  print(i)
  cluster.DE.GObp_summary <- rbind(
    cluster.DE.GObp_summary, 
    cluster.DE.GObp[[i]] %>% summary() %>% mutate(cluster = i-1)
    )
}


if(export){
cluster.DE.GObp_summary %>% write.csv2(file = paste0(source.output.dir, "/scRNA_cluster_GObp.csv"), row.names = FALSE)
}

cluster.DE.GObp_summary <- read.csv2(file = paste0(source.output.dir, "/scRNA_cluster_GObp.csv"))
cluster.DE.GObp_summary %>% split(cluster.DE.GObp_summary$cluster)
```

# sessionInfo - Package Versions
```{r}
sessionInfo()
# packageVersion("DESeq2")
# packinfo <- installed.packages(fields = c("Package", "Version"))
# packinfo[,c("Package", "Version")]
# packinfo["graphics",c("Package", "Version")]

```
