---
title: "Reproduction of Bakhoum et al 2018 Bulk Transcriptomics"
author: "Pedro Batista Tan"
date: "`r strftime(Sys.Date(), '%B %d %Y')`"
output:
  html_document:
    toc: yes
    number_sections: yes
  pdf_document:
    toc: yes
params:
  export: TRUE
---
# Install/load packages
```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, warning=FALSE, message = FALSE, dpi = 300, error = TRUE
  # fig.width = 10, fig.height = 6
)

# install.packages("BiocManager", "broom", "patchwork", "ggfortify", "colorspace", "GGally", "ggrepel", "AnnotationDbi", "tidyverse", "gplots", "pheatmap", "RColorBrewer", "msigdbr")
# BiocManager::install("DESeq2", "apeglm" "org.Hs.eg.db", "fgsea")
library(DESeq2)
library(broom)
library(patchwork)
library(ggfortify)
library(colorspace)
library(GGally)
library(ggrepel)
library(AnnotationDbi)
library(tidyverse)
library(gplots)
library(pheatmap)
library(RColorBrewer)
library(org.Hs.eg.db)
library(msigdbr)
library(fgsea)

select = dplyr::select

set.seed(42)
```

# Setup
```{r Setup, include=FALSE}
export = params$export
source.output.dir <- file.path("output_bulkPE75bp_sc5Samples")

if(export){
  # generate output dir path named data
  # if source output dir does not exist, create it
  if (!dir.exists(source.output.dir)) {
    dir.create(source.output.dir)
  } else{
    print("Output folder already exists")
  }
}
```

# Load data
```{r load_data, include=FALSE}
# Load transcriptomics count data
# Each row is a gene, each column a sample
df.counts.raw <- read.csv(file.path("data", "Bulk_data", "GSE98183_Counts.geneSymbols.75bpPE.csv"))

# load sample design
design <- read.csv2(file.path("data","Bulk_data","samples.csv"))
colnames(design) <- c("Sample", "Name", "Group")

# Sample names
colnames(df.counts.raw)
colnames(df.counts.raw) <- c("gene", "MK", "MKH", "Kb", "Ka", "cont")

df.counts.f <- df.counts.raw %>% column_to_rownames("gene")
df.counts.f <- df.counts.f[df.counts.f %>% rowMeans() > 0,]
```

# Map ensembl IDs to gene symbols
```{r}
# Some of the gene names in the count data correspond to the gene_name, some to the ENSEMBL gene_id
rownames(df.counts.f) %>% head(50)

gene_features <- read.csv2(file = file.path(dirname(getwd()), "biomaRt_gene_features.csv"))

gene_map <- rownames(df.counts.f) %>% as.data.frame() %>% rename("ensembl_gene_id" = ".") %>% filter(str_detect(ensembl_gene_id, "ENSG")) %>% left_join(gene_features %>% select(ensembl_gene_id, hgnc_symbol) %>% unique()) %>% unique()

# Two ensembl IDs have 2 different gene symbols
table(gene_map$ensembl_gene_id) %>% as.data.frame() %>% filter(Freq > 1)
gene_map %>% filter(ensembl_gene_id == "ENSG00000254876")
gene_map %>% filter(ensembl_gene_id == "ENSG00000276085")

# Choosing one of the 2 symbols for each (CCL3L1 and STRA6LP)
gene_map <- gene_map %>% filter(!hgnc_symbol %in% c("CCL3L3", "SUGT1P4-STRA6LP")) %>% unique()

gene_map %>% filter(is.na(hgnc_symbol)) %>% nrow()
gene_map %>% filter(hgnc_symbol == "") %>% nrow()

# genes with a mapped symbol for IDs
gene_map_f <- gene_map %>% filter(!is.na(hgnc_symbol)) %>% filter(hgnc_symbol != "")

```

```{r}
df.counts.symbol <- df.counts.f[!rownames(df.counts.f) %in% gene_map_f$ensembl_gene_id,]

# Replace ensembl ID for gene symbol, when available
df.counts.id <- df.counts.f[rownames(df.counts.f) %in% gene_map_f$ensembl_gene_id,]

df.counts.id <- df.counts.id %>% rownames_to_column("ensembl_gene_id") %>% left_join(gene_map_f, by = "ensembl_gene_id") %>% select(-ensembl_gene_id)

# Two IDs mapped to the same gene, #LINC01238
table(df.counts.id$hgnc_symbol) %>% as.data.frame() %>% filter(Freq > 1)

# Aggregate their counts
df.counts.id <- df.counts.id %>% group_by(hgnc_symbol) %>% summarize_at(colnames(df.counts.id)[!colnames(df.counts.id) %in% c("hgnc_symbol")], sum) %>% column_to_rownames("hgnc_symbol")

df.counts <- df.counts.id %>% rbind(df.counts.symbol)

rm(df.counts.id, df.counts.symbol)
```

```{r}
# Summary of dataframe with columns as factors
df.counts %>% lapply(as.character) %>% lapply(as.factor) %>% as.data.frame() %>% summary()
```

```{r}
# Compute and plot total counts per sample
df.counts %>% colSums() %>% data.frame() %>% rename("Value" = ".") %>% rownames_to_column("Sample") %>% 
  ggplot(aes(x = Sample, y = Value, fill = Sample)) + geom_col() + labs(x = "Sample_ID", y = "Raw Counts", title = "Raw total counts") + theme(plot.title = element_text(hjust= 0.5), legend.position = "none") + scale_y_continuous(labels = scales::comma) + scale_fill_viridis_d()

```

```{r}
# Convert dataframe to long format
df.counts.long <- df.counts %>% rownames_to_column("gene") %>% pivot_longer(cols = c(-gene), values_to = "Count",  names_to = "Sample")
```

# Filtering out low-count genes
```{r}
# Check how many genes have a sample with at least X counts
# df.counts.long %>%
#   group_by(gene) %>% filter(any(Count > 23)) %>% 
#   mutate(counts_log2 = log2(Count+1)) %>% ungroup() %>% select(gene) %>% n_distinct()

# filter out genes for which all the samples have 0 counts.
# Also create a column for log2 ofa pseudocounts
df.counts.long <- df.counts.long %>%
  group_by(gene) %>% filter(any(Count > 0)) %>% ungroup()

# Compare the number of genes
df.counts.raw$gene %>% n_distinct()
df.counts.long$gene %>% n_distinct()
```

# Visualizing count distribution
Visualize the distribution of the counts in log space.
```{r}
# make a boxplot using the counts_log2 column
df.counts.long %>% 
  group_by(gene) %>%  
  mutate(counts_log2 = log2(Count+1)) %>% ungroup() %>% ggplot(aes(x = Sample, y = counts_log2, col = Sample)) + geom_boxplot() + labs(x = "Sample", y = "log2 Counts", title = "Distribution of Log2 Counts") + theme(legend.position = "none", plot.title = element_text(hjust= 0.5)) + scale_fill_brewer(palette = "Dark2")

```

# Differential Expression
## Preparing the data for DESeq2
```{r}
# join design data
# Convert CIN-medium label to CIN-High, as done in the paper
design_c <- design %>% mutate(Group = ifelse(Group == "CIN-medium", "CIN-high", Group))

df.counts.long <- df.counts.long %>% left_join(design_c,  by = "Sample")
```

Create the object for countData. Transform the df.diff to wide format, set the genes as row names and convert to matrix.

```{r}
# select only the necessary columns
# tranform to wide format
# set the row names
# transform to to matrix.
df.diff.wide <- df.counts.long %>% 
  arrange(Group) %>% 
  select(gene, Count, Sample) %>% 
  pivot_wider(
    names_from = Sample, 
    values_from =  Count
    ) %>% column_to_rownames("gene") %>%
  as.matrix()

head(df.diff.wide)

# df.diff.wide <- df.counts.f %>% relocate(MK, .after = MKH) %>% as.matrix()
# df.counts.f[, c(2,1,3)]
```

```{r}
  # get the unique combinations of sample and group, arrange by group
  # convert Sample and group columns to factors.
  # Rows must be in the same order as the columns of the count matrix
df.diff.design <- df.counts.long %>% 
  distinct(Sample, Group) %>%
  arrange(Group) %>%
  mutate(
    Sample = as.factor(Sample),
    Group = as.factor(Group)
  )
head(df.diff.design)
head(df.diff.wide)
```

Create the DESeq2 object and estimate size factors to apply counts normalization

```{r}
# create the deseq dataset
deseq.dataset <- DESeqDataSetFromMatrix(
  countData=as.matrix(df.diff.wide),
  colData=df.diff.design,
  design=~Group
) %>%
  estimateSizeFactors()

# Change factor levels to compare CIN-High x CIN-low (default is alphabetical)
deseq.dataset$Group <- factor(deseq.dataset$Group, levels = c("CIN-low","CIN-high"))
```
## Dispersion estimation
```{r}
# run DESeq on dataset, create DESEq object
dds <- DESeq(deseq.dataset)

resultsNames(dds)
res <- results(dds)

summary(res)
```

```{r}
# Plot variance over mean of normalized counts to check overdispersion
norm.counts <- counts(deseq.dataset, normalized=TRUE)

df.mean.var <- tibble(
  mean_counts = rowMeans(norm.counts),
  var =  apply(norm.counts, 1, var)
)

df.mean.var %>% 
  ggplot(aes(x = mean_counts, y = var)) +
    geom_point(alpha = 0.2) +
  labs(
    title = "Mean-variance relationship",
    x = "Mean log2(normalized counts) per gene",
    y = "Variance of log2(normalized counts)"
  ) +
  geom_abline(intercept = 0, slope = 1, colour = "brown") +
  scale_x_log10() +
  scale_y_log10()

rm(df.mean.var)

plotDispEsts(dds, log = "xy")

```

## Differential expression
```{r}
df.deseq.results.noLFC <- res %>%
  as.data.frame() %>%
  rownames_to_column("gene")

df.deseq.results.shrunk <- dds %>%
  lfcShrink(coef = "Group_CIN.high_vs_CIN.low", type = "normal") %>%
  as.data.frame() %>%
  rownames_to_column("gene")

summary(df.deseq.results.shrunk)
```

```{r}
res %>% plotMA(ylim=c(-5,5))

# Check how many genes have an adj p-value < 0.01
df.deseq.results.noLFC %>% filter(padj < 0.01) %>% select(gene) %>% n_distinct()

# histogram of pvalues
df.deseq.results.noLFC %>% 
  ggplot(aes(pvalue)) + geom_histogram() +   labs(
    title = "DESeq2 p-value distribution", 
    x = "DESeq2 P-value", 
    y = "Number of genes"
  ) +  # you can add two ggplot objects with `+` with the patchwork package
  df.deseq.results.noLFC %>% 
  ggplot(aes(padj)) + geom_histogram() +   labs(
    title = "DESeq2 Adjusted p-value distribution", 
    x = "DESeq2 adjusted P-value", 
    y = "Number of genes"
  )
 # histogram of adj pvalues
```

## Volcano plots
```{r}
CIN_signature <- c('PELI2','BMP2','SHH','TNS4','RAB3B','ROBO1','ARHGAP28','CHN2','CST1','F13A1','CPVL','SEMA6D','NHSL2','GTF2IP7','DPYSL3','PCDH7','KHDRBS3','TRAC','TMEM156', 'CST4','CD24','FGF5','NTN4')

alpha = 1e-15
log2FCthresh = 3

df.deseq.results.noLFC %>% mutate(
  within_threshold = (padj < alpha & abs(log2FoldChange) > log2FCthresh)
) %>% ggplot(
  aes(x=log2FoldChange , y= -log10(padj + 1e-200), col = within_threshold)) +     geom_point(alpha = 0.5) +
geom_vline(xintercept = 0, col = "black") + 
geom_vline(xintercept = c(-log2FCthresh, log2FCthresh), col = "red") +   geom_hline(yintercept = -log10(alpha), colour = "red") + 
geom_text_repel(
  aes(x = log2FoldChange, y = -log10(padj + 1e-200), label = gene), data = df.deseq.results.noLFC %>% filter(gene %in% CIN_signature), colour = "grey20", min.segment.length = 0
  ) +
  theme(legend.position = "none", plot.title = element_text(hjust= 0.5)) +
  labs(
    title = "", 
    x = "Effect size: log2(fold-change)", 
    y = "-log10(adjusted p-value)"
  ) + ylim(0, 50) # + scale_x_continuous(breaks = seq(-6,9, by = 1))

if (export){
  ggsave(file = file.path(source.output.dir, "volcano_plot_strict.png"), width=5, height=4, dpi=300)
}

summary(df.deseq.results.shrunk)
df.deseq.results.shrunk %>% mutate(
  within_threshold = (padj < alpha & abs(log2FoldChange) > log2FCthresh)
) %>% ggplot(
  aes(x=log2FoldChange , y= -log10(padj + 1e-200), col = within_threshold)) + geom_point(alpha = 0.5) +
geom_vline(xintercept = 0, col = "black") + 
geom_vline(xintercept = c(-log2FCthresh, log2FCthresh), col = "red") +   geom_hline(yintercept = -log10(alpha), colour = "red") + 
geom_text_repel(
  aes(x = log2FoldChange, y = -log10(padj + 1e-200), label = gene), data = df.deseq.results.shrunk %>% filter(gene %in% CIN_signature), colour = "grey20", min.segment.length = 0
  ) +
  theme(legend.position = "none", plot.title = element_text(hjust= 0.5)) +
  labs(
    title = "", 
    x = "Effect size: log2(fold-change)", 
    y = "-log10(adjusted p-value)"
  ) + ylim(0, 50) # +   scale_x_continuous(breaks = seq(-6,9, by = 1))

if (export){
  ggsave(file = file.path(source.output.dir, "volcano_plot_shrunk.png"), width=5, height=4, dpi=300)
}

# Invert legend
df.deseq.results.shrunk %>% mutate(
  within_threshold = (padj < alpha & abs(log2FoldChange) > log2FCthresh)
) %>% ggplot(
  aes(x=log2FoldChange , y= -log10(padj + 1e-200), col = within_threshold)) + geom_point(alpha = 0.5) +
geom_vline(xintercept = 0, col = "black") + 
geom_vline(xintercept = c(-log2FCthresh, log2FCthresh), col = "red") +   geom_hline(yintercept = -log10(alpha), colour = "red") + 
geom_text_repel(
  aes(x = log2FoldChange, y = -log10(padj + 1e-200), label = gene), data = df.deseq.results.shrunk %>% filter(gene %in% CIN_signature), colour = "grey20", min.segment.length = 0
  ) +
  theme(legend.position = "none", plot.title = element_text(hjust= 0.5)) +
  labs(
    title = "", 
    x = "Effect size: log2(fold-change)", 
    y = "-log10(adjusted p-value)"
  ) + ylim(0, 50) + scale_color_manual(values=c("FALSE"="#00BFC4", "TRUE"="#F8766D"))
# +   scale_x_continuous(breaks = seq(-6,9, by = 1))

if (export){
  ggsave(file = file.path(source.output.dir, "volcano_plot_shrunk_c.png"), width=5, height=4, dpi=300)
}
```
## Export
```{r}
df.deseq.results.noLFC <- df.deseq.results.noLFC %>% mutate(neglog10padj = ifelse(padj != 0, -log10(padj), NA))  

df.deseq.results.shrunk <- df.deseq.results.shrunk %>% mutate(neglog10padj = ifelse(padj != 0, -log10(padj), NA))  

if (export){
  # Export data
  # Add a column for negative log10 of pvalues
  df.deseq.results.noLFC %>% write_csv2(file = file.path(source.output.dir, "bulk_transcriptomics_deseq_noLFC.csv"))
  
  df.deseq.results.shrunk %>% write_csv2(file = file.path(source.output.dir, "bulk_transcriptomics_deseq_shrunkLFC.csv"))
  
  saveRDS(dds, file.path(source.output.dir, "dds.rds"))
}

df.deseq.results.noLFC <- read_csv2(file = file.path(source.output.dir, "bulk_transcriptomics_deseq_noLFC.csv"))

df.deseq.results.shrunk <- read_csv2(file = file.path(source.output.dir, "bulk_transcriptomics_deseq_shrunkLFC.csv"))
```
# Load msigdb hallmarks gene set
```{r Load msigdb hallmarks}
msigdb_hallmarks_set <- read.csv2(file = file.path(dirname(getwd()), "msigdb_hallmarks_set.csv"))
msigdb_hallmarks_set <- split(msigdb_hallmarks_set$gene_symbol, msigdb_hallmarks_set$gs_name)

msigdb_hallmarks_names <- read.csv2(file =  file.path(dirname(getwd()), "names_hallmarks.csv"))

```

# FGSEA not shrunk LFC
```{r}
CIN_fc_rnk <- df.deseq.results.noLFC %>% select(gene, log2FoldChange) %>% deframe()

fgseaRes.noLFC <- fgsea(pathways = msigdb_hallmarks_set, 
                  stats    = CIN_fc_rnk,
                  minSize  = 15,
                  maxSize  = 500)

fgseaRes.noLFC <- fgseaRes.noLFC %>% mutate(within_threshold = padj < 0.01, "Enriched_in" = ifelse(NES > 0, "CIN-high", "CIN-low")) %>% left_join(msigdb_hallmarks_names)

fgseaRes.noLFC %>% ggplot(aes(x = log10(pval), y = NES)) +
  geom_point(aes(size = 0.5, alpha = 0.3, col = Enriched_in)) + 
  geom_hline(yintercept = 0, col = "black") +
  geom_text_repel(
  aes(x = log10(padj), y = NES, label = name), data = fgseaRes.noLFC %>% filter(within_threshold), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0, max.overlaps = 20
  ) +
  labs(x = "log10(pval)", y = "Normalized Enrichment Score", col = "Enriched in") + 
  scale_y_continuous(limits = c(-2,2), breaks = seq(-2,2,1)) + 
  scale_x_continuous(trans = "reverse") + scale_fill_manual(values = c("red","blue")) + guides(alpha = FALSE, size = FALSE)

if(export){
ggsave(file = file.path(source.output.dir,"fGSEA_bulk_results_fc_pval.png"), width=8, height=5, dpi=300)
}

fgseaRes.noLFC %>% ggplot(aes(x = log10(padj), y = NES)) +
  geom_point(aes(size = 0.5, alpha = 0.3, col = Enriched_in)) + 
  geom_hline(yintercept = 0, col = "black") +
  geom_text_repel(
  aes(x = log10(padj), y = NES, label = name), data = fgseaRes.noLFC %>% filter(within_threshold), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0, max.overlaps = 20
  ) +
  labs(x = "log10(FDRq)", y = "Normalized Enrichment Score", col = "Enriched in") + 
  scale_y_continuous(limits = c(-2,2), breaks = seq(-2,2,1)) + 
  scale_x_continuous(trans = "reverse") + scale_fill_manual(values = c("red","blue")) + guides(alpha = FALSE, size = FALSE)


if (export){
ggsave(file = file.path(source.output.dir,"fGSEA_bulk_results_fc.png"), width=8, height=5, dpi=300)

fgseaRes.noLFC %>% select(-leadingEdge) %>% write.csv2(file = file.path(source.output.dir,"bulk_fgsea_notshrunkfc_CIN_highlow.csv"), row.names = FALSE)
}

fgseaRes.noLFC <- read.csv2(file = file.path(source.output.dir,"bulk_fgsea_notshrunkfc_CIN_highlow.csv"))
```

# FGSEA not shrunk Z-Stat
```{r}
CIN_fc_rnk <- df.deseq.results.noLFC %>% select(gene, stat) %>% deframe()

fgseaRes.noLFC.stat <- fgsea(pathways = msigdb_hallmarks_set, 
                  stats    = CIN_fc_rnk,
                  minSize  = 15,
                  maxSize  = 500)

fgseaRes.noLFC.stat <- fgseaRes.noLFC.stat %>% mutate(within_threshold = padj < 0.01, "Enriched_in" = ifelse(NES > 0, "CIN-high", "CIN-low")) %>% left_join(msigdb_hallmarks_names)

# plotEnrichment(msigdb_hallmarks_set[["HALLMARK_KRAS_SIGNALING_UP"]],
#                CIN_fc_rnk) + labs(title="KRAS Signaling")
# 
# plotEnrichment(msigdb_hallmarks_set[["HALLMARK_INFLAMMATORY_RESPONSE"]],
#                CIN_fc_rnk) + labs(title="Inflammatory Response")
# 
# plotEnrichment(msigdb_hallmarks_set[["HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION"]],
#                CIN_fc_rnk) + labs(title="EMT")

fgseaRes.noLFC.stat %>% select(pval) %>% summary()

fgseaRes.noLFC.stat %>% ggplot(aes(x = log10(padj), y = NES)) +
  geom_point(aes(size = 0.5, alpha = 0.3, col = Enriched_in)) + 
  geom_hline(yintercept = 0, col = "black") +
  geom_text_repel(
  aes(x = log10(padj), y = NES, label = name), data = fgseaRes.noLFC.stat %>% filter(within_threshold | name %in% c("Apical Surface", "MYC Targets V1", "IFN-a Response", "IFN-y Response", "OxPhos", "TGF-B Signaling", "EMT", "IL2-STAT5 Signaling", "Wnt B-Catenin Signaling", "TNF-a Signaling via NKFB", "Inflammatory Response", "KRAS Signaling Up")), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0, max.overlaps = 20
  ) +
  labs(x = "log10(FDRq)", y = "Normalized Enrichment Score", col = "Enriched in") + 
  scale_y_continuous(limits = c(-3,3), breaks = seq(-3,3,1)) + 
  scale_x_continuous(trans = "reverse") + scale_fill_manual(values = c("red","blue")) + guides(alpha = FALSE, size = FALSE)


leading_edge <- fgseaRes.noLFC.stat %>% select(pathway, leadingEdge)

leading_edge_df <- setNames(data.frame(matrix(ncol = 3, nrow = 0)), c("pathway", "leadingedge_gene", "position"))

for (i in 1:length(leading_edge$leadingEdge)){
  pathway_leadingedge <- leading_edge$leadingEdge[[i]] %>% as.data.frame() %>% rename("leadingedge_gene"= ".") %>% rownames_to_column("position") %>% mutate(pathway = leading_edge$pathway[i]) %>% relocate(pathway, 1) %>% relocate(position, .after = leadingedge_gene)
  
  leading_edge_df <- leading_edge_df %>% rbind(pathway_leadingedge)
  rm(i, pathway_leadingedge)
}

if (export){
#png(filename = file.path(source.output.dir,"fGSEA_bulk_results_Zstat.png"), width=800, height=500)
ggsave(file = file.path(source.output.dir,"fGSEA_bulk_results_Zstat.png"), width=8, height=5, dpi=300)

fgseaRes.noLFC.stat %>% select(-leadingEdge) %>% write.csv2(file = file.path(source.output.dir,"bulk_fgsea_Zstat_CIN_highlow_3samples.csv"), row.names = FALSE)

leading_edge_df %>% write.csv2(file = file.path(source.output.dir, "fgsea_bulk_deseqstat_leadingedge_3samples.csv"), row.names = FALSE)
}

# fgseaRes.noLFC.stat <- read.csv2(file = file.path(source.output.dir,"bulk_fgsea_Zstat_CIN_highlow_3samples.csv"))

```
# FGSEA shrunk LFC
```{r}
CIN_fc_rnk_shrunk <- df.deseq.results.shrunk %>% select(gene, log2FoldChange) %>% deframe()

fgseaRes.shrunk <- fgsea(pathways = msigdb_hallmarks_set, 
                  stats    = CIN_fc_rnk_shrunk,
                  minSize  = 15,
                  maxSize  = 500)

fgseaRes.shrunk <- fgseaRes.shrunk %>% mutate(within_threshold = padj < 0.01, "Enriched_in" = ifelse(NES > 0, "CIN-high", "CIN-low")) %>% left_join(msigdb_hallmarks_names)

fgseaRes.shrunk %>% ggplot(aes(x = log10(pval), y = NES)) +
  geom_point(aes(size = 0.5, alpha = 0.3, col = Enriched_in)) + 
  geom_hline(yintercept = 0, col = "black") +
  geom_text_repel(
  aes(x = log10(pval), y = NES, label = name), data = fgseaRes.shrunk %>% filter(within_threshold | name %in% c("Apical Surface", "MYC Targets V1", "IFN-a Response", "IFN-y Response", "OxPhos", "TGF-B Signaling", "EMT", "IL2-STAT5 Signaling", "Wnt B-Catenin Signaling", "TNF-a Signaling via NKFB", "Inflammatory Response", "KRAS Signaling Up")), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0, max.overlaps = 20
  ) +
  labs(x = "log10(pval)", y = "Normalized Enrichment Score", col = "Enriched in") + 
  scale_y_continuous(limits = c(-2,2), breaks = seq(-2,2,1)) + 
  scale_x_continuous(trans = "reverse") + scale_fill_manual(values = c("red","blue")) + guides(alpha = FALSE, size = FALSE)

if(export){
ggsave(file = file.path(source.output.dir,"fGSEA_bulk_results_fc_pval.png"), width=7, height=5, dpi=300)
}

fgseaRes.shrunk %>% ggplot(aes(x = log10(padj), y = NES)) +
  geom_point(aes(size = 0.5, alpha = 0.3, col = Enriched_in)) + 
  geom_hline(yintercept = 0, col = "black") +
  geom_text_repel(
  aes(x = log10(padj), y = NES, label = name), data = fgseaRes.shrunk %>% filter(within_threshold), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0, max.overlaps = 20
  ) +
  labs(x = "log10(FDRq)", y = "Normalized Enrichment Score", col = "Enriched in") + 
  scale_y_continuous(limits = c(-2,2), breaks = seq(-2,2,1)) + 
  scale_x_continuous(trans = "reverse") + scale_fill_manual(values = c("red","blue")) + guides(alpha = FALSE, size = FALSE)


if (export){
ggsave(file = file.path(source.output.dir,"fGSEA_bulk_results_shrunkfc.png"), width=8, height=5, dpi=300)

fgseaRes.shrunk %>% select(-leadingEdge) %>% write.csv2(file = file.path(source.output.dir,"bulk_fgsea_shrunkfc_CIN_highlow.csv"), row.names = FALSE)
}

fgseaRes.shrunk <- read.csv2(file = file.path(source.output.dir,"bulk_fgsea_shrunkfc_CIN_highlow.csv"))

```

```{r}
fgseaRes.shrunk %>% mutate(padj = ifelse(padj < 0.001, 0.00101, padj)) %>% ggplot(aes(x = log10(padj), y = NES)) +
  geom_point(aes(size = 0.5, alpha = 0.3, col = Enriched_in)) + 
  geom_hline(yintercept = 0, col = "black") +
  geom_text_repel(
  aes(x = log10(padj), y = NES, label = name), data = fgseaRes.shrunk %>% mutate(padj = ifelse(padj < 0.001, 0.00101, padj)) %>% filter(within_threshold  | name %in% c("Apical Surface", "MYC Targets V1", "IFN-a Response", "IFN-y Response", "OxPhos", "TGF-B Signaling", "EMT", "IL2-STAT5 Signaling", "Wnt B-Catenin Signaling", "TNF-a Signaling via NKFB", "Inflammatory Response", "KRAS Signaling Up")), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0, max.overlaps = 20
  ) +
  labs(x = "FDRq", y = "Normalized Enrichment Score", col = "Enriched in") + 
  scale_y_continuous(limits = c(-2,2), breaks = seq(-2,2,1)) + 
  scale_x_continuous(trans = "reverse", limits = c(0, -3), breaks = c(-3, -2, -1, 0) , labels = c("<0.001", "0.01", "0.1", "1"))  + scale_fill_manual(values = c("red","blue")) + guides(alpha = FALSE, size = FALSE)

if (export){
ggsave(file = file.path(source.output.dir,"fGSEA_bulk_results_shrunk_p.png"), width=7, height=5, dpi=300)
}
```

# Original GSEA
```{r}
Bulk_gsea <- read.csv2(file.path("data", "Supp_table6.csv"))

Bulk_gsea <- Bulk_gsea %>% left_join(msigdb_hallmarks_names %>% rename(NAME = pathway)) %>% mutate("Enriched_in" = ifelse(NES > 0, "CIN-high", "CIN-low"))
		
Bulk_gsea %>% select(NES, FDR.q.val) %>% summary()

Bulk_gsea %>% mutate(FDR.q.val = ifelse(FDR.q.val < 0.001, 0.00101, FDR.q.val)) %>% ggplot(aes(x = log10(FDR.q.val), y = NES)) +
  geom_point(aes(size = 0.5, alpha = 0.3, col = Enriched_in)) + 
  geom_hline(yintercept = 0, col = "black") +
  geom_text_repel(
  aes(x = log10(FDR.q.val), y = NES, label = name), data = Bulk_gsea %>% mutate(FDR.q.val = ifelse(FDR.q.val < 0.001, 0.00101, FDR.q.val)) %>% filter(name %in% c("Apical Surface", "MYC Targets V1", "IFN-a Response", "IFN-y Response", "OxPhos", "TGF-B Signaling", "EMT", "IL2-STAT5 Signaling", "Wnt B-Catenin Signaling", "TNF-a Signaling via NKFB", "Inflammatory Response", "KRAS Signaling Up")), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0
  ) +
  labs(x = "log10(FDRq)", y = "Normalized Enrichment Score", col = "Enriched in") + 
  scale_y_continuous(limits = c(-2,2.5), breaks = seq(-2,2.5,1)) + 
  scale_x_continuous(trans = "reverse", limits = c(0, -3), breaks = c(-3, -2, -1, 0) , labels = c("<0.001", "0.01", "0.1", "1")) + scale_fill_manual(values = c("red","blue")) + guides(alpha = FALSE, size = FALSE)

```
# Compare Shrunk FC FGSEA and original GSEA NES
```{r}
# fGSEA x Bulk
fGSEA_comp <- Bulk_gsea %>% select(NAME, NES, FDR.q.val) %>% rename(orig_NES = NES, orig_FDR.q.val = FDR.q.val) %>% left_join(
  fgseaRes.shrunk %>% select(pathway, NES, pval, padj) %>% rename(NAME = pathway)
  ) %>% 
  left_join(
  msigdb_hallmarks_names %>% rename(NAME = pathway)
  ) %>% mutate(sign_agreement = sign(orig_NES) == sign(NES))

fGSEA_comp %>% ggplot(aes(x = orig_NES, y = NES, col = log10(padj), shape = padj < 0.2)) + 
  geom_point() + geom_abline(slope = 1) + 
  geom_text_repel(
    aes(x = orig_NES, y = NES, label = name), data = fGSEA_comp %>% filter(!sign_agreement), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0, max.overlaps = 20
  ) + labs(x = "Orig GSEA NES", y = "fGSEA NES", shape = "FDRq < 0.2", col = "log10(FDRq)") +
  scale_color_gradient2(low="black", mid = "blue", high="red", midpoint = -3)

if(export){
  ggsave(file=file.path(source.output.dir,"shrunkfGSEA_x_orig.png"), width=6, height=4, dpi=300)
}

# Compare GSEA FDRq values
fGSEA_comp %>% ggplot(aes(x=orig_FDR.q.val, y= padj, col = NES)) + geom_point() + geom_abline(slope = 1)  + labs(x = "orig_FDRq", y = "FDRq") +
  scale_color_gradient2(low="blue", mid = "white", high="red", midpoint = 0)

if(export){
  ggsave(file=file.path(source.output.dir,"shrunkfGSEA_x_orig,fdr.png"), width=6, height=4, dpi=300)
}
```

# Compare Z-stat FGSEA and original GSEA NES
```{r}
# fGSEA x Bulk
fGSEA_comp <- Bulk_gsea %>% select(NAME, NES, FDR.q.val) %>% rename(orig_NES = NES, orig_FDR.q.val = FDR.q.val) %>% left_join(
  fgseaRes.noLFC.stat %>% select(pathway, NES, pval, padj) %>% rename(NAME = pathway)
  ) %>% 
  left_join(
  msigdb_hallmarks_names %>% rename(NAME = pathway)
  ) %>% mutate(sign_agreement = sign(orig_NES) == sign(NES))

fGSEA_comp %>% ggplot(aes(x = orig_NES, y = NES, col = log10(padj), shape = padj < 0.2)) + 
  geom_point() + geom_abline(slope = 1) + 
  geom_text_repel(
    aes(x = orig_NES, y = NES, label = name), data = fGSEA_comp %>% filter(!sign_agreement), colour = "grey20", force = 3, force_pull = 2, min.segment.length = 0, max.overlaps = 20
  ) + labs(x = "Orig GSEA NES", y = "fGSEA NES", shape = "FDRq < 0.2", col = "log10(FDRq)") +
  scale_color_gradient2(low="black", mid = "blue", high="red", midpoint = -3)

if(export){
  ggsave(file=file.path(source.output.dir,"statfGSEA_x_orig.png"), width=6, height=4, dpi=300)
}

# Compare GSEA FDRq values
fGSEA_comp %>% ggplot(aes(x=orig_FDR.q.val, y= padj, col = NES)) + geom_point() + geom_abline(slope = 1)  + labs(x = "orig_FDRq", y = "FDRq") +
  scale_color_gradient2(low="blue", mid = "white", high="red", midpoint = 0)

if(export){
  ggsave(file=file.path(source.output.dir,"statfGSEA_x_orig,fdr.png"), width=6, height=4, dpi=300)
}
```

# Variance Stabilizing Transform
```{r}
# Apply a variance stabilizing transform
vsd <- vst(dds, blind=FALSE)
vsd %>% assay() %>% head()

```

```{r}
# Heatmap on VSD data using top n most expressed genes
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:200]
df <- as.data.frame(colData(dds)[,c("Sample","Group")])
pheatmap(assay(vsd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)

```

# Perform PCA on VST data
```{r}
PCAdata <-plotPCA(vsd, intgroup=c("Group"), returnData = TRUE)
# plotPCA(vsd, intgroup=c("Group"))
PCAdata

# plotPCA(vsd, intgroup=c("Group")) +
# geom_text_repel(
#   aes(x = PC1, y = PC2, label = name), data = plotPCA(vsd, intgroup=c("Group"), returnData = TRUE))

p <- plotPCA(vsd, intgroup=c("Group")) +
geom_text_repel(
  aes(x = PC1, y = PC2, label = name), show.legend = FALSE,
  data = plotPCA(vsd, intgroup=c("Group"), returnData = TRUE) %>% mutate(name = c("dnMCAK","Kif2a", "Control", "MCAK","Kif2b"))) + scale_color_manual(values=c("CIN-low"="#00BFC4", "CIN-high"="#F8766D"))

p

if(export){
ggsave(plot = p, file = file.path(source.output.dir,"vst_pca.png"), width=4, height=4, dpi=300)
  
PCAdata %>% write_csv2(file = file.path(source.output.dir,"vst_pca_data.csv"))
}
```

```{r}
df.deseq.results.noLFC <- read_csv2(file = file.path(source.output.dir, "bulk_transcriptomics_deseq_noLFC.csv"))
```

# Comparing CIN signature
```{r}
CIN_signature <- c('PELI2','BMP2','SHH','TNS4','RAB3B','ROBO1','ARHGAP28','CHN2','CST1','F13A1','CPVL','SEMA6D','NHSL2','GTF2IP7','DPYSL3','PCDH7','KHDRBS3','TRAC','TMEM156', 'CST4','CD24','FGF5','NTN4')

# All genes are present
CIN_signature %in% df.deseq.results_map$gene

# Genes in volcano plot quadrant
df.deseq.results.noLFC %>% ggplot(
  aes(x=log2FoldChange , y= -log10(padj + 1e-200), col = within_threshold)) +     geom_point(alpha = 0.5) +
geom_vline(xintercept = 0, col = "black") + 
geom_vline(xintercept = c(-log2FCthresh, log2FCthresh), col = "red") +   geom_hline(yintercept = -log10(alpha), colour = "red") + 
geom_text_repel(
  aes(x = log2FoldChange, y = -log10(padj + 1e-200), label = gene), data = df.deseq.results.noLFC %>% filter(symbol %in% CIN_signature), colour = "grey20", min.segment.length = 0, max.overlaps = 20
  ) +
  theme(legend.position = "none", plot.title = element_text(hjust= 0.5)) +
  labs(
    title = "Volcano plot", 
    x = "Effect size: log2(fold-change)", 
    y = "-log10(adjusted p-value)"
  ) + ylim(0, 50) + 
  scale_x_continuous(breaks = seq(-6,9, by = 1))

if (export){
  ggsave(file = file.path(source.output.dir, "volcano_plot_noLFC_CIN_signature.png"), width=6, height=4, dpi=300)
}
```

# Original data
```{r}
df.deseq.results.orig <- read_excel(file.path("data", "orig_CINHighvsLow.xlsx"))

colnames(deseq.norm.orig)
colnames(deseq.norm.orig) <- c("gene","MK_orig","Kb_orig","MKH_orig","cont_orig","Ka_orig")

column_names <- colnames(df.deseq.results.orig)
# column_names <- replace(column_names, column_names == "...1", "gene")

column_names <- paste0("orig_", column_names)
column_names <- replace(column_names, column_names == "orig_...1", "gene")

colnames(df.deseq.results.orig) <- column_names
```

```{r}
deseq.noLFC.c <- df.deseq.results.noLFC %>% left_join(df.deseq.results.orig, by = "gene")

deseq.noLFC.c %>% ggplot(aes(x = log2FoldChange, y = orig_log2FoldChange)) + geom_point() + geom_abline(slope = 1)

```

